## Imports

In [48]:
import pandas as pd
import xgboost as xgb
from sklearn.model_selection import train_test_split
from sklearn.metrics import f1_score
from sklearn.model_selection import GridSearchCV

In [2]:
# load the data
pca_data = pd.read_csv('PCA_data.csv')

In [3]:
pca_data

Artist                             Song  Popular  \
0                      Yung Bleu                Life Worth Living        0   
1                      Yung Bleu                Everytime I Blink        0   
2                      Yung Bleu                             Good        0   
3                      Yung Bleu                    Come By At 12        0   
4                      Yung Bleu                   No, I'm Not Ok        0   
...                          ...                              ...      ...   
1762                  The Weeknd                   Take My Breath        1   
1763                  The Weeknd                        Heartless        1   
1764  YoungBoy Never Broke Again                           Bandit        1   
1765                Mariah Carey  All I Want For Christmas Is You        1   
1766                       Nelly                          Lil Bit        1   

      priciple_component_1  priciple_component_2  priciple_component_3  \
0                -1.362949              4.462221              5.261660   
1                 9.734700              4.760353             -6.760258   
2                 2.971137              3.878585              1.574890   
3                12.912365             -0.973002             -0.813052   
4                13.159638             -1.614713             -0.330539   
...                    ...                   ...                   ...   
1762              1.378909              2.652400            -11.799436   
1763             -5.455050              5.723004              1.111635   
1764             -4.760963             10.770093              5.451684   
1765              0.652380             -9.778914             -1.485794   
1766             -5.281438              2.913535              8.049488   

      priciple_component_4  priciple_component_5  priciple_component_6  \
0                -2.187603              4.844853              0.574108   
1                -3.193606              1.584642             -3.693776   
2                -0.324684             -6.988592             -6.166271   
3                -1.635890             -8.997236              0.348971   
4                 7.278265             -6.654534              1.748100   
...                    ...                   ...                   ...   
1762             -2.613351              3.295676              0.071900   
1763              5.163507              2.910969             -5.538088   
1764             -0.633699              2.715087             -1.721272   
1765             -1.998266             -0.645282             -7.116893   
1766             -3.401303             -2.381396             -1.992816   

      priciple_component_7  ...  priciple_component_162  \
0                 4.188373  ...               -0.246064   
1                -4.163491  ...               -0.627916   
2                 2.200268  ...               -0.159936   
3                 2.021174  ...               -0.684971   
4                -1.990139  ...                0.312657   
...                    ...  ...                     ...   
1762              2.136736  ...               -0.980382   
1763             -1.981216  ...                0.168967   
1764             -1.724169  ...                0.753133   
1765             -2.580430  ...                0.430934   
1766             -4.060882  ...               -0.650404   

      priciple_component_163  priciple_component_164  priciple_component_165  \
0                   0.582953                0.365841                0.604060   
1                   0.393207               -0.580349               -0.152202   
2                   0.486991                0.632565                0.826580   
3                  -0.843556                2.023126               -0.590720   
4                   1.090296               -0.006741                0.743082   
...                      ...                     ...                     ...   
1762               -1.188214              

In [4]:
pca_data.describe()

Popular  priciple_component_1  priciple_component_2  \
count  1767.000000          1.767000e+03          1.767000e+03   
mean      0.286927          2.412709e-16          9.650835e-17   
std       0.452455          7.745278e+00          6.817848e+00   
min       0.000000         -2.294522e+01         -1.572559e+01   
25%       0.000000         -5.538791e+00         -5.564498e+00   
50%       0.000000         -6.516017e-01         -4.517231e-01   
75%       1.000000          5.088034e+00          5.124766e+00   
max       1.000000          2.300332e+01          2.012141e+01   

       priciple_component_3  priciple_component_4  priciple_component_5  \
count          1.767000e+03           1767.000000          1.767000e+03   
mean           1.286778e-16              0.000000          1.688896e-16   
std            5.315700e+00              4.559876          4.478081e+00   
min           -1.528338e+01            -12.104846         -1.104701e+01   
25%           -3.379511e+00             -3.207507         -3.201926e+00   
50%           -7.817044e-02             -0.326280         -7.946891e-01   
75%            3.256416e+00              3.246575          3.146769e+00   
max            2.144773e+01             14.124034          1.533023e+01   

       priciple_component_6  priciple_component_7  priciple_component_8  \
count          1.767000e+03          1.767000e+03          1.767000e+03   
mean           1.125931e-16         -4.825418e-17         -6.433890e-17   
std            4.133736e+00          3.551619e+00          3.302479e+00   
min           -1.287771e+01         -1.088489e+01         -9.729959e+00   
25%           -2.719923e+00         -2.386988e+00         -2.242288e+00   
50%           -2.839046e-01         -1.700760e-01         -4.435594e-02   
75%            2.340314e+00          2.340240e+00          2.044797e+00   
max            1.679257e+01          1.429001e+01          1.144579e+01   

       priciple_component_9  ...  priciple_component_162  \
count          1.767000e+03  ...            1.767000e+03   
mean           6.433890e-17  ...           -1.396858e-15   
std            3.156061e+00  ...            6.256552e-01   
min           -1.124553e+01  ...           -2.145592e+00   
25%           -2.024711e+00  ...           -4.385592e-01   
50%           -1.288827e-01  ...           -9.162553e-03   
75%            1.950095e+00  ...            4.197647e-01   
max            1.253710e+01  ...            2.292284e+00   

       priciple_component_163  priciple_component_164  priciple_component_165  \
count            1.767000e+03            1.767000e+03            1.767000e+03   
mean            -4.659544e-16            6.041825e-16            1.648684e-16   
std              6.217794e-01            6.174367e-01            6.109375e-01   
min             -1.912519e+00           -2.186375e+00           -2.229110e+00   
25%             -4.027861e-01           -4.279048e-01           -4.156067e-01   
50%             -1.426289e-02           -1.333455e-02           -1.664720e-03   
75%              4.025537e-01            4.339853e-01            4.022737e-01   
max              2.179267e+00            2.421397e+00            2.292188e+00   

       priciple_component_166  priciple_component_167  priciple_component_168  \
count            1.767000e+03            1.767000e+03            1.767000e+03   
mean             1.125931e-16           -2.935462e-16           -2.513238e-16   
std              6.080542e-01            6.013372e-01            6.001581e-01   
min             -2.015239e+00           -2.131600e+00           -2.090494e+00   
25%             -4.027813e-01           -4.068708e-01           -3.841582e-01   
50%              6.141814e-03           -1.028217e-02           -2.394802e-02   
75%              4.069481e-01            4.051081e-01            3.794563e-01   
max              2.245335e+00            2.344646e+00            2.217012e+00   

       priciple_component_169  priciple_component_170  priciple

In [5]:
pca_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1767 entries, 0 to 1766
Columns: 174 entries, Artist to priciple_component_171
dtypes: float64(171), int64(1), object(2)
memory usage: 2.3+ MB


## Data Modeling with XGBoost

In [6]:
# remove unnecessary columns
pca_data = pca_data.drop(columns=['Artist', 'Song'])

In [9]:
# divide data into X and y
X = pca_data.drop('Popular', axis=1)
y = pca_data[['Popular']]

In [13]:
# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [32]:
# Define the XGBoost model
xgb_model = xgb.XGBClassifier()

In [42]:
xgb_param_grid = {
    'learning_rate': [0.01,0.1,0.5,0.9],
    'n_estimators': [150, 250, 400, 500],
    'subsample': [0.3, 0.5, 0.9],
    'max_depth': [5, 7, 10]
}

In [43]:
xgb_grid = GridSearchCV(
    estimator= xgb_model, 
    param_grid= xgb_param_grid, 
    scoring= 'f1',
    cv= 4, 
    verbose= 1
)

In [44]:
xgb_grid.fit(X, y)

Fitting 4 folds for each of 144 candidates, totalling 576 fits


GridSearchCV(cv=4,
             estimator=XGBClassifier(base_score=None, booster=None,
                                     callbacks=None, colsample_bylevel=None,
                                     colsample_bynode=None,
                                     colsample_bytree=None,
                                     early_stopping_rounds=None,
                                     enable_categorical=False, eval_metric=None,
                                     feature_types=None, gamma=None,
                                     gpu_id=None, grow_policy=None,
                                     importance_type=None,
                                     interaction_constraints=None,
                                     learning_rate=None,...
                                     max_delta_step=None, max_depth=None,
                                     max_leaves=None, min_child_weight=None,
                                     missing=nan, monotone_constraints=None,
                                     n_estimators=100, n_jobs=None,
                                     num_parallel_tree=None,
                                     objective='reg:logistic', predictor=None, ...),
             param_grid={'learning_rate': [0.01, 0.1, 0.5, 0.9],
                         'max_depth': [5, 7, 10],
                         'n_estimators': [150, 250, 400, 500],
                         'subsample': [0.3, 0.5, 0.9]},
             scoring='f1', verbose=1)

In [46]:
print("Best parameters found: ", xgb_grid.best_params_)
print("Best F1-score found: ", xgb_grid.best_score_)

Best parameters found:  {'learning_rate': 0.1, 'max_depth': 5, 'n_estimators': 500, 'subsample': 0.9}
Best F1-score found:  0.5672876308249382


In [49]:
# Define the XGBoost model with optimal hyperparameters
xgb_tuned = xgb.XGBClassifier(learning_rate=0.1, max_depth=5, n_estimators=500, subsample=0.9)

In [50]:
# Fit the model on the training data
xgb_tuned.fit(X_train, y_train)

XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric=None, feature_types=None,
              gamma=None, gpu_id=None, grow_policy=None, importance_type=None,
              interaction_constraints=None, learning_rate=0.1, max_bin=None,
              max_cat_threshold=None, max_cat_to_onehot=None,
              max_delta_step=None, max_depth=5, max_leaves=None,
              min_child_weight=None, missing=nan, monotone_constraints=None,
              n_estimators=500, n_jobs=None, num_parallel_tree=None,
              predictor=None, random_state=None, ...)

In [51]:
# perform prediction using XGBoost tuned model
y_pred = xgb_tuned.predict(X_test)

In [52]:
# Evaluate the F1 score of the XGBoost tuned model
score = f1_score(y_test, y_pred)
print('F1 Score: %.3f' % score)

F1 Score: 0.571


So after tuning the hyperparameters, we can see that the best parameters are:
1. learning_rate = 0.1
2. max_depth= 5
3. n_estimators= 500
4. subsample= 0.9

We have seen that there was improvement in the F1 score after tuning the hyperparameters. and the current F1 score is 0.571